### PR TITLE
MTL PSM2: add a thread lock while peeking and completing the psm2 requests.

### PR DESCRIPTION
**Pull request by matcabral**
MTL PSM2: add a thread lock while peeking and completing the psm2 requests.
requests.
Reviewed-by: Gopalakrishnan, Aravind <aravind.gopalakrishnan@intel.com>
Signed-off-by: Matias Cabral <matias.a.cabral@intel.com>

### CODE REVIEW

**Comment by matcabral** [comment]
solves #3880

File: ompi/mca/mtl/psm2/mtl_psm2_component.c [hunk 1]
```diff
@@ -361,6 +365,8 @@ ompi_mtl_psm2_component_init(bool enable_progress_threads,
     int verno_minor = PSM2_VERNO_MINOR;
     int local_rank = -1, num_local_procs = 0;
 
+    OBJ_CONSTRUCT(&mtl_psm2_mq_mutex, opal_mutex_t);
```
Comment (rhc54) [inline ompi/mca/mtl/psm2/mtl_psm2_component.c]: you did the static init above, so I don't believe you need to construct it again here

File: ompi/mca/mtl/psm2/mtl_psm2.c [hunk 2]
```diff
@@ -406,58 +406,62 @@ int ompi_mtl_psm2_progress( void ) {
...
 
     do {
+        OPAL_THREAD_LOCK(&mtl_psm2_mq_mutex);
         err = psm2_mq_ipeek2(ompi_mtl_psm2.mq, &req, NULL);
-	if (err == PSM2_MQ_INCOMPLETE) {
-	    return completed;
-	} else if (err != PSM2_OK) {
-	    goto error;
-	}
+        if (err == PSM2_MQ_INCOMPLETE) {
```
Comment (rhc54) [inline ompi/mca/mtl/psm2/mtl_psm2.c]: is this likely or unlikely? or does it matter? I noted you used opal_unlikely below, hence the question

Comment (matcabral) [reply ompi/mca/mtl/psm2/mtl_psm2.c]: mm... both should happen in every call. The loop will iterate until emptying PSM2's  Matched Queue.  How many times will go on each direction will depend on each app, e.g. in flight yet to be poled/tested requests.

File: ompi/mca/mtl/psm2/mtl_psm2_component.c (hunk 1 above)
Comment (matcabral) [reply ompi/mca/mtl/psm2/mtl_psm2_component.c]: Ok, will remove and re-test. OBJ_DESTRUCT will also fly away.

File: ompi/mca/mtl/psm2/mtl_psm2.c (hunk 2 above)
Comment (matcabral) [reply ompi/mca/mtl/psm2/mtl_psm2.c]: oh, the unlikely is for psm2_mq_test2() failing to on a psm request that psm2_mq_ipeek2() returned as ready to be tested. That should never fail, but..

**Comment by matcabral** [comment]
removed the obj creation and destruction

**Comment by rhc54** [comment]
github failure

bot:ompi:retest

**Comment by matcabral** [comment]
bot:ompi:retest

**Comment by matcabral** [comment]
Hi, 
Not sure what "Pull Request Build Checker — Build triggered. sha1 is merged."  means. Can this actually be merged? thanks,

**Comment by rhc54** [comment]
The tester was stuck for most of the day yesterday - let's restart it

bot:ompi:retest

**Comment by rhc54** [comment]
@bwbarrett Is the PR checker stuck? It seems to have been hanging here for quite some time.

**Comment by rhc54** [comment]
one more try before we punt

bot:ompi:retest

**Comment by matcabral** [comment]
thanks @rhc54

**Pull request merged**